pandas/core/generic.py

```diff
@@ -5439,11 +5439,7 @@ def _dir_additions(self) -> Set[str]:
         add the string-like attributes from the info_axis.
         If info_axis is a MultiIndex, it's first level values are used.
         """
-        additions = {
-            c
-            for c in self._info_axis.unique(level=0)[:100]
-            if isinstance(c, str) and c.isidentifier()
-        }
+        additions = self._info_axis._dir_additions_for_owner
```

Comment:
instead of adding all of this code, why can't you slap a `cache_readonly` here?

Reply:
That's not possible, because the axes (index/columns) can be changed. However, the axis labels are immutable, so it's safe to use `cache_readonly` there.

I've changed the PR to avoid the addition in child classes.